File: src/openrtm_aist/src/lib/rtm/NVUtil.cpp
// -*- C++ -*-
/*!
 * @file NVUtil.cpp
 * @brief NameValue and NVList utility functions
 * @date $Date: 2008-01-13 07:41:08 $
 * @author Noriaki Ando <n-ando@aist.go.jp>
 *
 * Copyright (C) 2006-2008
 *     Noriaki Ando
 *     Task-intelligence Research Group,
 *     Intelligent Systems Research Institute,
 *     National Institute of
 *         Advanced Industrial Science and Technology (AIST), Japan
 *     All rights reserved.
 *
 * $Id$
 *
 */

#ifdef WIN32
#pragma warning( push )
#pragma warning( disable : 4267 )
#pragma warning( disable : 4290 )
#pragma warning( disable : 4311 )
#pragma warning( disable : 4312 )
#endif // WIN32

#include <map>
#include <algorithm>
#include <coil/stringutil.h>
#include <rtm/NVUtil.h>
#include <rtm/CORBA_SeqUtil.h>

#ifdef WIN32
#pragma warning( pop )
#endif // WIN32

namespace NVUtil
{
  /*!
   * @if jp
   * @brief value が CORBA::Char の NameValue を生成する
   * @else
   * @brief Create NameValue typed CORBA::Char
   * @endif
   */
  SDOPackage::NameValue newNVChar(const char* name, const CORBA::Char value)
  {
    SDOPackage::NameValue nv;
    nv.name = CORBA::string_dup(name);
    nv.value <<= CORBA::Any::from_char(value);
    return nv;
  }
  
  /*!
   * @if jp
   * @brief value が CORBA::Boolean の NameValue を生成する
   * @else
   * @brief This operation creates NameValue typed CORBA::Boolean.
   * @endif
   */
  SDOPackage::NameValue newNVBool(const char* name, const CORBA::Boolean value)
  {
    SDOPackage::NameValue nv;
    nv.name = CORBA::string_dup(name);
    nv.value <<= CORBA::Any::from_boolean(value);
    return nv;
  }
  
  /*!
   * @if jp
   * @brief value が CORBA::Octet の NameValue を生成する
   * @else
   * @brief Create NameValue typed CORBA::Octet
   * @endif
   */
  SDOPackage::NameValue newNVOctet(const char* name, const CORBA::Octet value)
  {
    SDOPackage::NameValue nv;
    nv.name = CORBA::string_dup(name);
    nv.value <<= CORBA::Any::from_octet(value);
    return nv;
  }
  
  /*!
   * @if jp
   * @brief value が CORBA::Any の NameValue を生成する
   * @else
   * @brief Create NameValue typed CORBA::Any
   * @endif
   */
  SDOPackage::NameValue newNVAny(const char* name, const CORBA::Any& value)
  {
    SDOPackage::NameValue nv;
    nv.name = CORBA::string_dup(name);
    nv.value = value;
    return nv;
  }
  
  /*!
   * @if jp
   * @brief Properties を NVList へコピーする
   * @else
   * @brief Copy the properties to NVList
   * @endif
   */
#ifndef ORB_IS_RTORB
  void copyFromProperties(SDOPackage::NVList& nv, const coil::Properties& prop)
#else // ORB_IS_RTORB
    void copyFromProperties(SDOPackage_NVList& nv, const coil::Properties& prop)
#endif // ORB_IS_RTORB
  {
    std::vector<std::string> keys;
    keys = prop.propertyNames();
    CORBA::ULong len((CORBA::ULong)keys.size());
    nv.length(len);
    
    for (CORBA::ULong i = 0; i < len; ++i)
      {
// Why RtORB does not copy string to Properties.
#ifndef ORB_IS_RTORB
        nv[i].name = CORBA::string_dup(keys[i].c_str());
#else // ORB_IS_RTORB
        nv[i].name = (char *)keys[i].c_str();
#endif // ORB_IS_RTORB
	nv[i].value <<= prop[keys[i]].c_str();
      }
  }
  
  /*!
   * @if jp
   * @brief NVList を Properties へコピーする
   * @else
   * @brief Copy NVList to the Proeprties
   * @endif
   */
  void copyToProperties(coil::Properties& prop, const SDOPackage::NVList& nv)
  {
    for (CORBA::ULong i(0), len(nv.length()); i < len; ++i)
      {
	const char* value;
	if (nv[i].value >>= value)
	  {
	    const char* name(nv[i].name);
	    prop[name] = value;
	  };
      }
  }
  
  /*!
   * @if jp
   * @brief NVList を Properties に変換するためのファンクタ
   * @else
   * @brief Functor to transform NVList into the properties
   * @endif
   */
  struct to_prop
  {
    to_prop()
    {
    };
    void operator()(const SDOPackage::NameValue& nv)
    {
      const char* value;
      if (nv.value >>= value)
	{
	  m_prop.setProperty(CORBA::string_dup(nv.name), value);
	};
    }
    coil::Properties m_prop;
  };
  
  /*!
   * @if jp
   * @brief NVList を Properties へ変換する
   * @else
   * @brief Transform NVList to the properties
   * @endif
   */
  coil::Properties toProperties(const SDOPackage::NVList& nv)
  {
    to_prop p;
    p = CORBA_SeqUtil::for_each(nv, p);
    return p.m_prop;
  }
  
  /*!
   * @if jp
   * @brief NVList を検索するためのファンクタ
   * @else
   * @brief Functor to find a NVList
   * @endif
   */
  struct nv_find
  {
    nv_find(const char* name) : m_name(name) {};
    bool operator()(const SDOPackage::NameValue& nv)
    {
      std::string name(nv.name);
      return m_name == name;
    }
    std::string m_name;
  };
  
  /*!
   * @if jp
   * @brief NVList から name で指定された value を返す
   * @else
   * @brief Return the value specified by name from NVList
   * @endif
   */
  const CORBA::Any& find(const SDOPackage::NVList& nv, const char* name)
  {
    CORBA::Long index;
    index = CORBA_SeqUtil::find(nv, NVUtil::nv_find(name));
    if (index < 0) throw std::string("Not found");
    return nv[index].value;
  }
  
  /*!
   * @if jp
   * @brief name で指定された要素のインデックスを返す
   * @else
   * @brief Return the index of element specified by name from NVList
   * @endif
   */
  const CORBA::Long find_index(const SDOPackage::NVList& nv, const char* name)
  {
    return  CORBA_SeqUtil::find(nv, NVUtil::nv_find(name));
  }
  
  /*!
   * @if jp
   * @brief 指定された name の value の型が string であるか検証する
   * @else
   * @brief Validate whether value type specified by name is string type
   * @endif
   */
  bool isString(const SDOPackage::NVList& nv, const char* name)
  {
    try
      {
	CORBA::Any value;
	value = find(nv, name);
	const char* str_value;
	return value >>= str_value;
      }
    catch (...)
      {
	return false;
      }
  }
  
  /*!
   * @if jp
   * @brief 指定された name の value の値が指定した文字列と一致するか検証する
   * @else
   * @brief Check whether the value of specified name specified matches
   *        the specified string
   * @endif
   */
  bool isStringValue(const SDOPackage::NVList& nv,
		     const char* name, const char* value)
  {
    if (isString(nv, name))
      {
	if (toString(nv, name) == value)
	  {
	    return true;
	  }
      }
    return false;
  }
  
  /*!
   * @if jp
   * @brief 指定された name の NVList を string として返す。
   * @else
   * @brief Get NVList of specifid name as string
   * @endif
   */
  std::string toString(const SDOPackage::NVList& nv, const char* name)
  {
    const char* str_value;
    try
      {
	if(!(find(nv, name) >>= str_value))
          {
	    str_value = "";
          }
      }
    catch (...)
      {
	str_value = "";
      }

    if (str_value == NULL)
      {
	str_value = "";
      }
    
    return str_value;
  }
  
  /*!
   * @if jp
   * @brief 指定された文字列を NVList の要素に追加する。
   * @else
   * @brief Append the specified string to element of NVList
   * @endif
   */
#ifndef ORB_IS_RTORB
  bool appendStringValue(SDOPackage::NVList& nv, const char* name,
                         const char* value)
#else // ORB_IS_RTORB
  bool appendStringValue(SDOPackage_NVList& nv, const char* name,
                         const char* value)
#endif // ORB_IS_RTORB
  {
    //    if (!isString(nv, name)) return false;
    
    CORBA::Long index;
    index = find_index(nv, name);
    
    if (index < 0)
      {
	CORBA_SeqUtil::push_back(nv, newNV(name, value));
      }
    else
      {
	const char* tmp_char;
	nv[index].value >>= tmp_char;
	std::string tmp_str(tmp_char);
	
	std::vector<std::string> values;
	values = coil::split(tmp_str, ",");
	if (values.end() == std::find(values.begin(), values.end(), value))
	  {
	    tmp_str.append(",");
	    tmp_str.append(value);
	    nv[index].value <<= tmp_str.c_str();
	  }
      }
    return true;
  }
  
  /*!
   * @if jp
   * @brief NVList に要素を追加する。
   * @else
   * @brief Append an element to NVList
   * @endif
   */
  void append(SDOPackage::NVList& dest, const SDOPackage::NVList& src)
  {
    for (CORBA::ULong i = 0, len = src.length(); i < len; ++i)
      {
	CORBA_SeqUtil::push_back(dest, src[i]);
      }
  }
  
  /*!
   * @if jp
   * @brief NVList に設定されている内容を文字列として出力する。
   * @else
   * @brief Print information configured in NVList as a string type
   * @endif
   */
  std::ostream& dump_to_stream(std::ostream& out, const SDOPackage::NVList& nv)
  {
    for (CORBA::ULong i(0), n(nv.length()); i < n; ++i)
      {
	const char* str_value;
	if (nv[i].value >>= str_value)
	  {
	    out << nv[i].name << ": " << str_value << std::endl;
	  }
	else
	  {
	    out << nv[i].name << ": not a string value" << std::endl;
	  }
      }
    return out;
  }

  /*!
   * @if jp
   * @brief NVList に設定されている内容を文字列として標準出力する。
   * @else
   * @brief Print information configured in NVList as a string type 
   *        to Standard Outport.
   * @endif
   */
  void dump(const SDOPackage::NVList& nv)
  {
    dump_to_stream(std::cout, nv);
  }

  /*!
   * @if jp
   * @brief NVList に設定されている内容を文字列にする
   * @else
   * @brief Get information configured in NVList as a string type
   * @endif
   */
  std::string toString(const SDOPackage::NVList& nv)
  {
    std::stringstream s;
    dump_to_stream(s, nv);
    return s.str();
  }
};

File: src/openrtm_aist_python/OpenRTM_aist/NVUtil.py
#!/usr/bin/env python
# -*- coding: euc-jp -*- 

##
# @file NVUtil.py
# @brief NameValue and NVList utility functions
# @date $Date: 2007/09/11$
# @author Noriaki Ando <n-ando@aist.go.jp> and Shinji Kurihara
# 
# Copyright (C) 2006-2008
#     Noriaki Ando
#     Task-intelligence Research Group,
#     Intelligent Systems Research Institute,
#     National Institute of
#         Advanced Industrial Science and Technology (AIST), Japan
#     All rights reserved.

import sys
import traceback
from omniORB import any

import OpenRTM_aist
import SDOPackage, SDOPackage__POA


##
# @if jp
#
# @brief NameValue を生成する
#
# このオペレーションはNameValueを作成する。
#
# @param name NameValue の name
# @param value NameValue の value
#
# @return NameValue
#
# @else
#
# @brief Create NameVale
#
# This operation creates NameVale.
#
# @param name name of NameValue
# @param value value of NameValue
#
# @return NameValue
#
# @endif
def newNV(name, value):
  try:
    any_val = any.to_any(value)
  except:
    print "ERROR  NVUtil.newNV : Can't convert to any. ", type(value)
    raise

    
  nv = SDOPackage.NameValue(name, any_val)
  return nv


##
# @if jp
#
# @brief Properties を NVList へコピーする
#
# このオペレーションは Properties を NVList へコピーする。
# NVList の value は全て CORBA::string 型としてコピーする。
#
# @param nv Properties の値を格納する NVList
# @param prop コピー元の Properties
#
# @else
#
# @brief Copy to NVList from Proeprties
#
# This operation copies Properties to NVList.
# Created NVList's values are CORBA::string.
#
# @param nv NVList to store Properties values
# @param prop Properties that is copies from
#
# @endif
# void copyFromProperties(SDOPackage::NVList& nv, const coil::Properties& prop);
def copyFromProperties(nv, prop):
  keys = prop.propertyNames()
  keys_len = len(keys)
  nv_len = len(nv)
  if nv_len > 0:
    for i in range(nv_len):
      del nv[-1]

  for i in range(keys_len):
    nv.append(newNV(keys[i], prop.getProperty(keys[i])))


##
# @if jp
#
# @brief NVList を Properties へコピーする
#
# このオペレーションは NVList を Properties へコピーする。
#
# @param prop NVList の値を格納する Properties
# @param nv コピー元の NVList
#
# @else
#
# @brief Copy to Proeprties from NVList
#
# This operation copies NVList to Properties.
#
# @param prop Properties to store NVList values
# @param nv NVList that is copies from
#
# @endif
# void copyToProperties(coil::Properties& prop, const SDOPackage::NVList& nv);
def copyToProperties(prop, nvlist):
  for nv in nvlist:
    try:
      val = str(any.from_any(nv.value, keep_structs=True))
      prop.setProperty(str(nv.name),val)
    except:
      print OpenRTM_aist.Logger.print_exception()
      pass



##
# @if jp
# @class to_prop
# @brief NVList → Properties 変換用ファンクタ
# @endif
class to_prop:
  def __init__(self):
    self._prop = OpenRTM_aist.Properties()
    
  def __call__(self, nv):
    self._prop.setProperty(nv.name, nv.value)



##
# @if jp
#
# @brief NVList を Properties へ変換する
#
# このオペレーションは NVList を Properties へ変換する。
#
# @param nv 変換元の NVList
#
# @return 変換結果Property
#
# @else
#
# @endif
# coil::Properties toProperties(const SDOPackage::NVList& nv);
def toProperties(nv):
  p = OpenRTM_aist.CORBA_SeqUtil.for_each(nv, to_prop())
  return p._prop



##
# @if jp
# @class nv_find
# @brief NVList 検索用ファンクタ
# @endif
class nv_find:
  """
  """

  def __init__(self, name):
    self._name = name

  def __call__(self, nv):
    return str(self._name) == str(nv.name)


##
# @if jp
#
# @brief NVList から name で指定された value を返す
#
# このオペレーションは name で指定された value を Any 型で返す。
# 指定した名称の要素が存在しない場合は例外を発生させる。
#
# @param nv 検索対象の NVList
# @param name 検索する名前
#
# @return 検索結果
#
# @else
#
# @brief Get value in NVList specified by name
#
# This operation returns Any type of value specified by name.
# Created NVList's values are CORBA::string.
#
# @param nv NVList to be searched
# @param prop name to seartch in NVList
#
# @endif
def find(nv, name):
  index = OpenRTM_aist.CORBA_SeqUtil.find(nv, nv_find(name))

  if index < 0:
    raise "Not found."

  return nv[index].value


##
# @if jp
#
# @brief name で指定された要素のインデックスを返す
#
# このオペレーションは name で指定された要素が格納されている位置の
# インデックスを返す。
#
# @param nv 検索対象の NVList
# @param name 検索する名前
#
# @return 検索対象のインデックス
#
# @else
#
# @endif
def find_index(nv, name):
  return OpenRTM_aist.CORBA_SeqUtil.find(nv, nv_find(name))


##
# @if jp
#
# @brief 指定された name の value の型が string であるか検証する
#
# このオペレーションは name で指定された value の型が CORBA::string
# かどうかを bool 値で返す。
#
# @param nv 検索対象の NVList
# @param name 検索する名前
#
# @return string検証結果(string:true、それ以外:false)
#
# @else
#
# @endif
def isString(nv, name):
  try:
    value = find(nv, name)
    val = any.from_any(value, keep_structs=True)
    return type(val) == str
  except:
    return False


##
# @if jp
#
# @brief 指定された name の value の型が指定した文字列と一致するか検証する
#
# このオペレーションは name で指定された value の型が CORBA::string
# かどうかを判断し、  CORBA::string である場合には指定した文字列と一致するか
# をbool 値で返す。
#
# @param nv 検索対象の NVList
# @param name 検索する名前
# @param value 比較対象文字列
#
# @return 検証結果(文字列と一致:true、非一致:false)
#
# @else
#
# @endif
def isStringValue(nv, name, value):
  if isString(nv, name):
    if toString(nv, name) == value:
      return True
  return False


##
# @if jp
#
# @brief 指定された name の NVList を string として返す。
#
# このオペレーションは name で指定された NVList の値を string で返す。
# もし、name で指定した value の値が CORBA::string でなければ、
# 空の文字列のstringを返す。
#
# @param nv 検索対象の NVList
# @param name 検索する名前
#
# @return name に対応する値のstring型の値
#
# @else
#
# @brief Get string value in NVList specified by name
#
# This operation returns string value in NVList specified by name.
# If the value in NVList specified by name is not CORBA::string type
# this operation returns empty string value.
#
# @param nv NVList to be searched
# @param name name to to serach
#
# @return string value named by name
#
# @endif
def toString(nv, name=None):
  if not name:
    str_ = [""]
    return dump_to_stream(str_, nv)

  str_value = ""
  try:
    ret_value = find(nv, name)
    val = any.from_any(ret_value, keep_structs=True)
    if type(val) == str:
      str_value = val
  except:
    print OpenRTM_aist.Logger.print_exception()
    pass
  
  return str_value


##
# @if jp
#
# @brief 指定された文字列を NVList の要素に追加する。
#
# このオペレーションは name で指定された要素に value で指定された文字列を
# 追加する。
# name で指定した要素に既に value の値が設定されている場合には何もしない。
# name で指定した要素に value の値が設定されていない場合は、 ｢,｣区切りで
# value の値を追加する。
# 指定された値を設定する。
# name で指定した要素が存在しない場合は、 NVList の最後に新たな要素を追加し、
# 指定された値を設定する。
#
# @param nv 検索対象の NVList
# @param name 追加対象要素名
# @param value 追加する文字列
#
# @return 追加操作結果
#
# @else
#
# @endif
def appendStringValue(nv, name, value):
  index = find_index(nv, name)
  if index >= 0:
    tmp_str = nv[index].value.value()
    values = OpenRTM_aist.split(tmp_str,",")
    find_flag = False
    for val in values:
      if val == value:
        find_flag = True

    if not find_flag:
      tmp_str += ", "
      tmp_str += value
      nv[index].value = any.to_any(tmp_str)
  else:
    OpenRTM_aist.CORBA_SeqUtil.push_back(nv, newNV(name, value))

  return True


##
# @if jp
#
# @brief NVList に要素を追加する。
#
# このオペレーションは dest で指定された NVList に src で指定された要素を
# 追加する。
#
# @param dest 追加される NVList
# @param src 追加する NVList
#
# @else
#
# @endif
def append(dest, src):
  for i in range(len(src)):
    OpenRTM_aist.CORBA_SeqUtil.push_back(dest, src[i])


##
# @if jp
# @brief NVList に設定されている内容を文字列として出力する。
# @else
# @brief Print information configured in NVList as a string type
# @endif
# std::ostream& dump_to_stream(std::ostream& out, const SDOPackage::NVList& nv)
def dump_to_stream(out, nv):
  for i in range(len(nv)):
    val = any.from_any(nv[i].value, keep_structs=True)
    if type(val) == str:
	    out[0] += (nv[i].name + ": " + str(nv[i].value) + "\n")
    else:
	    out[0] += (nv[i].name + ": not a string value \n")

  return out[0]


##
# @if jp
#
# @brief NVList に設定されている内容を文字列として出力する。
#
# 指定された NVList に設定された内容を文字列として出力する。
# なお、設定されている要素が文字列型以外の場合には、その旨(文字列ではない)を
# 出力する。
#
# @param nv 出力対象 NVList
#
# @else
#
# @endif
def dump(nv):
  out = [""]
  print dump_to_stream(out, nv)
